R/checks.R

```diff
@@ -57,51 +58,60 @@ check_user_rights <- function(
 #' \code{REDCapR::redcap_read_oneshot()$data}
 #'
 #' @importFrom rlang .data
-#' @importFrom dplyr slice pull %>% select
-#' @importFrom tidyr drop_na
-#' @importFrom tidyselect everything
+#' @importFrom dplyr %>% select mutate case_when
+#' @importFrom purrr map2
 #'
 #' @keywords internal
 
 
 check_repeat_and_nonrepeat <- function(db_data) {
 
-  # Loop through dataframe columns then rows to determine if a variable has data
-  # and appears alongside redcap_repeat_instance when it is both NA and not NA
-
-  #start at 5 to skip record id, redcap event, instrument, and instance vars
-  for (i in 5:length(names(db_data))) {
-
-    nonrepeating <- NA
-    repeating <- NA
-
-    for (j in seq_len(nrow(db_data))){
+  # This check function looks for potential repeat/nonrepeat behavior using the
+  # steps below:
+  # 1) Define standard columns that don't need checking and remove those from
+  #    analysis (i.e. "safe columns").
+  # 2) Create a dummy column for each remaining column and use case_when to
+  #    to assign whether the column demonstrates repeating, nonrepeating, or
+  #    indeterminant behavior per row. Indetemrinant would be the case for
+  #    data not yet filled out.
+  # 3) Use a mapping function to determine if any dummy columns contain both
+  #    "repeating" AND "nonrepeating" declarations, if so error out.
+
+  # Step (1)
+  safe_cols <- names(db_data)[1:4]
```

Comment:
Is this true that the first 4 columns of `db_data` are always safe? What if the database has arms or no arms?

Reply:
Ah that's a good point, I think I can correct this to look for the first 3 instead when `redcap_event_name` is missing i.e. `is_longitudinal` is T/F and reorder things.

Reply:
Sounds good, resolving using callout and change from `all_of` to `any_of`, pushing commit now:

```
  safe_cols <- c(names(db_data)[1], "redcap_event_name",
                 "redcap_repeat_instrument", "redcap_repeat_instance")

  # Step (2)
  check_data <- db_data %>%
    mutate(
      across(.cols = -any_of(safe_cols),
             .names = "{.col}_repeatingcheck",
     ........
```